src/cluster_legacy.c

```diff
@@ -4319,26 +4358,72 @@ void clusterSendUpdate(clusterLink *link, clusterNode *node) {
+ * If is_light is 1, then build a message block with `clusterMsgLight` struct else `clusterMsg`. */
+static clusterMsgSendBlock *createModuleMsgBlock(int64_t module_id, uint8_t type, const char *payload, uint32_t len, int is_light) {
+    uint32_t msglen;
+    int msgtype;
+    clusterMsgSendBlock *msgblock;
+    clusterMsgModule *module_data;
+
+    if (is_light) {
+        msglen = sizeof(clusterMsgLight) - sizeof(union clusterMsgData);
+        msgtype = CLUSTERMSG_TYPE_MODULE | CLUSTERMSG_LIGHT;
+    } else {
+        msglen = sizeof(clusterMsg) - sizeof(union clusterMsgData);
+        msgtype = CLUSTERMSG_TYPE_MODULE;
+    }
     msglen += sizeof(clusterMsgModule) - 3 + len;
-    clusterMsgSendBlock *msgblock = createClusterMsgSendBlock(CLUSTERMSG_TYPE_MODULE, msglen);
+    msgblock = createClusterMsgSendBlock(msgtype, msglen);
 
-    clusterMsg *hdr = getMessageFromSendBlock(msgblock);
-    hdr->data.module.msg.module_id = module_id; /* Already endian adjusted. */
-    hdr->data.module.msg.type = type;
-    hdr->data.module.msg.len = htonl(len);
-    memcpy(hdr->data.module.msg.bulk_data, payload, len);
+    if (is_light) {
+        clusterMsgLight *hdr = getLightMessageFromSendBlock(msgblock);
+        module_data = &hdr->data.module.msg;
+    } else {
+        clusterMsg *hdr = getMessageFromSendBlock(msgblock);
+        module_data = &hdr->data.module.msg;
+    }
 
-    if (link)
-        clusterSendMessage(link, msgblock);
-    else
-        clusterBroadcastMessage(msgblock);
+    module_data->module_id = module_id; /* Already endian adjusted */
+    module_data->type = type;
+    module_data->len = htonl(len);
+    memcpy(module_data->bulk_data, payload, len);
 
-    clusterMsgSendBlockDecrRefCount(msgblock);
+    return msgblock;
+}
+
+/* Send a MODULE message.
+ *
+ * If link is NULL, then the message is broadcasted to the whole cluster. */
+void clusterSendModule(clusterLink *link, uint64_t module_id, uint8_t type, const char *payload, uint32_t len) {
+    clusterMsgSendBlock *msgblock = NULL, *msgblock_light = NULL;
+    ClusterNodeIterator iter;
+
+    if (link) {
+        clusterNodeIterNode(&iter, link->node);
+    } else {
+        /* Broadcast to all the nodes. */
+        clusterNodeIterInitAllNodes(&iter);
+    }
+    clusterNode *node;
+    while ((node = clusterNodeIterNext(&iter)) != NULL) {
+        if (node->flags & (CLUSTER_NODE_MYSELF | CLUSTER_NODE_HANDSHAKE)) continue;
+        if (nodeSupportsLightMsgHdrForModule(node)) {
+            if (msgblock_light == NULL) {
+                msgblock_light = createModuleMsgBlock(module_id, type, payload, len, 1);
+            }
+            clusterSendMessage(node->link, msgblock_light);
+        } else {
+            if (msgblock == NULL) {
+                msgblock = createModuleMsgBlock(module_id, type, payload, len, 0);
+            }
+            clusterSendMessage(node->link, msgblock);
+        }
```

Comment:
Sorry for diving into these details. It's not very important but I can't help seeing the possibility to de-duplicate some more here. Something like this? (untested)

```suggestion
        int is_light = nodeSupportsLightMsgHdrForModule(node);
        clusterMsgSendBlock **block = is_light ? &msgblock_light : &msgblock;
        if (*block == NULL) {
            *block = createModuleMsgBlock(module_id, type, payload, len, is_light);
        }
        clusterSendMessage(node->link, *block);
```

Even more compact (but maybe more obscure?) is to use an array `clusterMsgSendBlock *msgblock[2]` and use `is_light` as an index into that array.

```c
        int is_light = nodeSupportsLightMsgHdrForModule(node);
        if (msgblock[is_light] == NULL) {
            msgblock[is_light] = createModuleMsgBlock(module_id, type, payload, len, is_light);
        }
        clusterSendMessage(node->link, msgblock[is_light]);
```

Is it better? Idk, you can decide.

Reply:
Looks clean. Incorporated.